utils/zerovec/src/map/hashmap.rs

```diff
@@ -0,0 +1,287 @@
+        let mut reverse_mapping = vec![0; len];
+
+        let mut start = 0;
+        while start < len {
+            // Bucket span with the same first level hash
+            // start is always within bounds of `bucket_flatten`
+            let g = bucket_flatten[start].0 .0;
+            // g is always within bounds of `bucket_sizes`
+            let end = start + bucket_sizes[g];
+            // start, end - 1 are always within bounds of `bucket_sizes`
+            let buckets = &bucket_flatten[start..end];
+
+            'd0: for d0 in 0..len as u32 {
+                'd1: for d1 in 0..len as u32 {
+                    current_displacements.clear();
+                    generation += 1;
+
+                    for ((_, f0, f1), _) in buckets {
+                        let displacement_idx =
+                            compute_displacement((*f0, *f1), (d0, d1), len as u32);
+
+                        // displacement_idx is always within bounds
+                        if occupied[displacement_idx] || assignments[displacement_idx] == generation
+                        {
+                            continue 'd1;
+                        }
+                        assignments[displacement_idx] = generation;
+
+                        current_displacements.push(displacement_idx);
+                    }
+
+                    // Successfully found a (d0, d1), store it as index g.
+                    // g < displacements.len() due to modulo operation
+                    displacements[g] = (d0, d1);
+
+                    for (i, displacement_idx) in current_displacements.iter().enumerate() {
+                        // `current_displacements` has same size as `buckets`
+                        let (_, idx) = &buckets[i];
+
+                        // displacement_idx is always within bounds
+                        occupied[*displacement_idx] = true;
+                        reverse_mapping[*displacement_idx] = *idx;
+                    }
+                    break 'd0;
+                }
+            }
+
+            start = end;
+        }
+
+        keys.zvl_permute(&mut reverse_mapping.clone());
+        values.zvl_permute(&mut reverse_mapping);
+
+        Self {
+            displacements: ZeroVec::alloc_from_slice(&displacements),
+        }
+    }
+
+    #[inline]
+    pub fn index<K>(&self, k: &K) -> Option<usize>
+    where
+        K: Hash,
+    {
+        let (g, f0, f1) = compute_hash(k, self.displacements.len());
+        let (d0, d1) = self.displacements.get(g).unwrap();
+        Some(compute_displacement(
+            (f0, f1),
+            (d0, d1),
+            self.displacements.len() as u32,
+        ))
+    }
+}
+
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct ZeroHashMapStatic<'a, K, V>
+where
+    K: ZeroMapKV<'a> + ?Sized,
+    V: ZeroMapKV<'a> + ?Sized,
+{
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    index: HashIndex<'a>,
```

Comment:
I don't see the advantage of splitting `HashIndex` into a different type. This makes both construction and indexing spread out over two types/functions, which is harder to follow.